langserver/references.go

```diff
@@ -447,6 +465,241 @@ func findReferences(ctx context.Context, lconf loader.Config, pkgInWorkspace fun
+	// It comes at the cost of some code complexity.
+	//
+	// Here's a high level summary.
+	//
+	// The goal is to find references to the query object p.Q.
+	// There are several possible scenarios, each handled differently.
+	//
+	// 1. We are looking in a package other than p, and p is not dot-imported.
+	//    This is the simplest case. Q must be referred to as n.Q,
+	//    where n is the name under which p is imported.
+	//    We look at all imports of p to gather all names under which it is imported.
+	//    (In the typical case, it is imported only once, under its default name.)
+	//    Then we look at all selector expressions and report any matches.
+	//
+	// 2. We are looking in a package other than p, and p is dot-imported.
+	//    In this case, Q will be referred to just as Q.
+	//    Furthermore, go/ast's object resolution will not be able to resolve
+	//    Q to any other object, unlike any local (file- or function- or block-scoped) object.
+	//    So we look at all matching identifiers and report all unresolvable ones.
+	//
+	// 3. We are looking in package p.
+	//    (Care must be taken to separate p and p_test (an xtest package),
+	//    and make sure that they are treated as separate packages.)
+	//    In this case, we give go/ast the entire package for object resolution,
+	//    instead of going file by file.
+	//    We then iterate over all identifiers that resolve to the query object.
+	//    (The query object itself has already been reported, so we don't re-report it.)
+	//
+	// We always skip all files that don't contain the string Q, as they cannot be
+	// relevant to finding references to Q.
+	//
+	// We parse all files leniently. In the presence of parsing errors, results are best-effort.
+
+	defpkg := strings.TrimSuffix(obj.Pkg().Path(), "_test")
+	defpkg = util.VendorlessImportPath(defpkg)
+
+	defname := obj.Pkg().Name()                    // name of the defining package of the query object, used for resolving imports that use import path only (common case)
+	isxtest := strings.HasSuffix(defname, "_test") // indicates where the query object is defined in an xtest package
+
+	name := obj.Name()
+	namebytes := []byte(name)          // byte slice version of query object name, for early filtering
+	objpos := fset.Position(obj.Pos()) // position of query object, used to prevent re-emitting original decl
+
+	var files []string    // reusable list of files
+	var pkgnames []string // reusable list of names the package is imported under
+
+	find := h.getFindPackageFunc()
+
+	for u := range users {
+		// Bail out early if the context is canceled
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		uIsXTest := strings.HasSuffix(u, "!test") // indicates whether this package is the special defpkg xtest package
+		u = strings.TrimSuffix(u, "!test")
+
+		if !pkgInWorkspace(u) {
+			continue
+		}
+
+		// Resolve package.
+		// TODO: is fromDir == "" correct?
+		pkg, err := find(ctx, bctx, u, "", build.IgnoreVendor)
```

Comment:
`fromDir` can be empty if we don't have relative import paths (which I don't think we should since users is from the importgraph?) https://sourcegraph.com/github.com/golang/go@go1.9/-/blob/src/go/build/build.go#L527